Source/Core/Core/HW/DVDThread.cpp

```diff
@@ -0,0 +1,188 @@
+static u64 s_realtime_started_us;
+static u64 s_realtime_done_us;
+
+void Start()
+{
+	s_finish_read = CoreTiming::RegisterEvent("FinishReadDVDThread", FinishRead);
+	_assert_(!s_DVD_thread.joinable());
+	s_DVD_thread = std::thread(DVDThread);
+}
+
+void Stop()
+{
+	_assert_(s_DVD_thread.joinable());
+
+	// The DVD thread will return if s_DVD_thread_exiting
+	// is set when it starts working
+	s_DVD_thread_exiting.Set();
+	s_DVD_thread_start_working.Set();
+
+	s_DVD_thread.join();
+
+	s_DVD_thread_exiting.Clear();
+}
+
+void DoState(PointerWrap &p)
+{
+	WaitUntilIdle();
+
+	// TODO: Savestates can be smaller if s_DVD_buffer is not saved
+	p.Do(s_DVD_buffer);
+	p.Do(s_time_read_started);
+	p.Do(s_DVD_success);
+
+	p.Do(s_DVD_offset);
+	p.Do(s_output_address);
+	p.Do(s_length);
+	p.Do(s_decrypt);
+	p.Do(s_callback_event_type);
+
+	// s_realtime_started_us and s_realtime_done_us aren't savestated
+	// because they rely on the current system's time.
+	// This means that loading a savestate might cause
+	// incorrect times to be logged once.
+}
+
+void WaitUntilIdle()
+{
+	_assert_(Core::IsCPUThread());
+
+	// Wait until DVD thread isn't working
+	s_DVD_thread_not_working.Wait();
+
+	// Set the event again so that we still know that the DVD thread isn't working
+	s_DVD_thread_not_working.Set();
+}
+
+void StartRead(u64 DVD_offset, u32 output_address, u32 length, bool decrypt,
+               int callback_event_type, int ticks_until_completion)
+{
+	_assert_(Core::IsCPUThread());
+
+	s_DVD_thread_not_working.Wait();
+
+	s_DVD_offset = DVD_offset;
+	s_output_address = output_address;
+	s_length = length;
+	s_decrypt = decrypt;
+	s_callback_event_type = callback_event_type;
+
+	s_time_read_started = CoreTiming::GetTicks();
+	s_realtime_started_us = Common::Timer::GetTimeUs();
+
+	s_DVD_thread_start_working.Set();
+
+	CoreTiming::ScheduleEvent(ticks_until_completion, s_finish_read);
+}
+
+static void FinishRead(u64 userdata, int cyclesLate)
+{
+	WaitUntilIdle();
+
+	DEBUG_LOG(DVDINTERFACE, "Disc has been read. Real time: %llu us. Real time including delay: %llu us. Emulated time including delay: %llu us.",
+	          s_realtime_done_us - s_realtime_started_us,
+	          Common::Timer::GetTimeUs() - s_realtime_started_us,
+	          (CoreTiming::GetTicks() - s_time_read_started) / (SystemTimers::GetTicksPerSecond() / 1000 / 1000));
+
+	while (!s_DVD_success)
+	{
+		// Asking "Do you want to retry?" would be more user friendly,
+		// but if panic alerts are turned off, the answer is automatically Yes,
+		// and we don't want an infinite loop where ReadFromDisc always fails...
+		if (PanicYesNoT("The disc could not be read (at 0x%" PRIx64 " - 0x%" PRIx64 ").\n"
+		                "Do you want to continue without retrying?",
+		                s_DVD_offset, s_DVD_offset + s_length))
+			break;
+		else
+			s_DVD_success = ReadFromDisc();
+	}
+
+	if (s_DVD_success)
+		Memory::CopyToEmu(s_output_address, s_DVD_buffer.data(), s_length);
```

Comment:
We really need it to be deterministic over anything else. Determinism isn't just useful for netplay; Bugs are hard enough to find when they are reproducible, it gets a lot harder when timings can vary based on on your computer's setup or other factors.

The DVD could start writing data instantly if the data was already in the OS's cache, potentially before the CPU thread even continues execution. Or it could hit a spinning HDD and take 10-12ms.

Maybe if we had locking (hint, hint me) we could lock a region of memory and allow other threads to write into main memory while maintaining determinism.